ElectronicObserver/Window/Tools/DropRecordViewer/DropRecordViewerViewModel.cs

```diff
@@ -349,341 +360,292 @@ private string GetMapString(int serialID, bool insertEnemyFleetName = true)
-		{
-			DropRecordOption.All => 0,
-			DropRecordOption.Drop => 1,
-			_ => 2
-		};
+	[RelayCommand(IncludeCancelCommand = true)]
+	private async Task Search(CancellationToken cancellationToken)
+	{
+		SearchStartTime = DateTime.UtcNow;
+		StatusInfoText = EncycloRes.SearchingNow;
 
-		int priorityContent = Math.Max(priorityShip, priorityItem);
+		try
+		{
+			if (MergeRows)
+			{
+				IEnumerable<MergedDropRecordRow> rows = await Task.Run(MakeMergedDropRecordRows, cancellationToken);
 
-		List<ShipDropRecord.ShipDropElement> records = RecordManager.Instance.ShipDrop.Record;
-		LinkedList<DropRecordRow> rows = new();
+				MergedRecordRows = new(rows);
+				DataGridMergedRowsViewModel.ItemsSource = MergedRecordRows;
+			}
+			else
+			{
+				IEnumerable<DropRecordRow> rows = await Task.Run(MakeDropRecordRows, cancellationToken);
 
+				RecordRows = new(rows);
+				DataGridRawRowsViewModel.ItemsSource = RecordRows;
+			}
 
-		//lock ( records )
+			StatusInfoText = $"{EncycloRes.SearchComplete} ({(int)(DateTime.UtcNow - SearchStartTime).TotalMilliseconds} ms)";
+		}
+		catch (OperationCanceledException)
 		{
-			int i = 0;
-			var counts = new Dictionary<string, int[]>();
-			var allcounts = new Dictionary<string, int[]>();
-
+			StatusInfoText = EncycloRes.SearchCancelled;
+		}
+		catch (Exception e)
+		{
+			// this should never happen?
+			Logger.Add(2, "", e);
+		}
+	}
 
-			foreach (var r in records)
-			{
+	[SuppressMessage("Critical Code Smell", "S3776:Cognitive Complexity of methods should not be too high", Justification = "<Pending>")]
+	private bool ShouldIncludeInMergedCount(ShipDropRecord.ShipDropElement r)
+	{
+		if (r.Date < DateTimeBegin || DateTimeEnd < r.Date) return false;
+		if (!RankS && r.Rank is "SS" or "S") return false;
+		if (!RankA && r.Rank is "A") return false;
+		if (!RankB && r.Rank is "B") return false;
+		if (!RankX && Constants.GetWinRank(r.Rank) <= BattleRank.C) return false;
 
-				#region Filtering
+		if (MapAreaID is int world && world != r.MapAreaID) return false;
+		if (MapInfoID is int map && map != r.MapInfoID) return false;
+		if (MapCellID.Ids?.Contains(r.CellID) == false) return false;
 
-				if (r.Date < DateTimeBegin || DateTimeEnd < r.Date)
-					continue;
+		if (IsBossOnly is false && r.IsBossNode) return false;
+		if (IsBossOnly is true && !r.IsBossNode) return false;
 
-				if (((r.Rank == "SS" || r.Rank == "S") && !RankS) ||
-					((r.Rank == "A") && !RankA) ||
-					((r.Rank == "B") && !RankB) ||
-					((Constants.GetWinRank(r.Rank) <= 3) && !RankX))
-					continue;
+		if (MapDifficulty is int difficulty && difficulty != r.Difficulty) return false;
 
+		return true;
+	}
 
-				if (MapAreaID is int world && world != r.MapAreaID) continue;
-				if (MapInfoID is int map && map != r.MapInfoID) continue;
-				if (MapCellID.Ids?.Contains(r.CellID) == false) continue;
+	[SuppressMessage("Critical Code Smell", "S3776:Cognitive Complexity of methods should not be too high", Justification = "<Pending>")]
```

Comment:
i think you can remove the cognitive complexity warning with return early

Reply:
nope
![image](https://github.com/ElectronicObserverEN/ElectronicObserver/assets/40002167/148b59d3-3f7d-4025-879f-18425691d792)